### PR TITLE
Some fancy new mocha tests for List Exports

### PR DESCRIPTION
**Pull request by biyeun**
@benrudolph @czue

(Download Export tests are coming)

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 50 errors:

* corehq/apps/export/static/export/spec/data/list_exports.data.mo.js, line 52 - Missing semicolon.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 1 - Use the function form of "use strict".
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 166 - Missing semicolon.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 3 - 'describe' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 11 - 'beforeEach' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 12 - 'angular' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 31 - 'module' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 33 - 'inject' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 36 - 'beforeEach' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 36 - 'inject' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 47 - 'describe' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 49 - 'afterEach' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 54 - 'it' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 59 - 'ListExportsTestData' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 62 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 63 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 66 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 67 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 70 - 'it' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 78 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 79 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 82 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 83 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 84 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 87 - 'it' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 94 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 95 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 96 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 101 - 'describe' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 103 - 'beforeEach' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 109 - 'ListExportsTestData' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 110 - 'ListExportsTestData' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 111 - 'ListExportsTestData' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 116 - 'afterEach' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 121 - 'it' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 126 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 127 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 129 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 130 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 133 - 'it' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 138 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 139 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 141 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 142 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 144 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 145 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 148 - 'it' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 161 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 163 - 'assert' is not defined.
* corehq/apps/export/static/export/spec/list_exports.spec.mo.js, line 164 - 'assert' is not defined.

File: corehq/apps/export/static/export/spec/list_exports.spec.mo.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            var exportToUpdate = currentScope.exports[0];
+            var component = exportToUpdate.emailedExports[0];
+            $httpBackend.expectPOST(mockBackendUrls.UPDATE_EMAILED_EXPORT_DATA);
+            currentScope.updateEmailedExportData(component, exportToUpdate);
+            assert.isTrue(component.updatingData);
+            $httpBackend.flush();
+            assert.isFalse(component.updatingData);
+            assert.isTrue(component.updatedDataTriggered);
+        });
+    })
```
Comment (dimagimon) [inline corehq/apps/export/static/export/spec/list_exports.spec.mo.js]: Missing semicolon.

File: corehq/apps/export/static/export/spec/list_exports.spec.mo.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    describe('Initialization', function() {
+
+        afterEach(function() {
+            $httpBackend.verifyNoOutstandingExpectation();
+            $httpBackend.verifyNoOutstandingRequest();
+        });
+
+        it('registers a non-blank list of exports', function() {
+            $httpBackend
```
Comment (benrudolph) [inline corehq/apps/export/static/export/spec/list_exports.spec.mo.js]: another way to mock out the backend that doesn't require modifying the existing code is to use sinon's fake server functionality: http://sinonjs.org/docs/#fakeServer

Comment (benrudolph) [inline corehq/apps/export/static/export/spec/list_exports.spec.mo.js]: although it looks like angular has its own framework for mocking http requests?

File: corehq/apps/export/templates/export/spec/mocha.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+{% load hq_shared_tags %}
+
+{% block core_libraries %}{{ block.super }}
+{% include "style/includes/bootstrap3/angular.html" %}
+<script src="{% new_static 'angular-mocks/angular-mocks.js' %}"></script>
+{% endblock %}
+
+{% block dependencies %}
+<script src="{% new_static 'export/js/list_exports.ng.js' %}"></script>
+{#<script type="text/javascript">#}
```
Comment (benrudolph) [inline corehq/apps/export/templates/export/spec/mocha.html]: is this commented out? sorry i forget what `{#` signifies

File: corehq/apps/mocha/README.md [hunk 4]
```diff
@@ -1,6 +1,33 @@
 Mocha Tests
 ===========
 
+## Getting started
```
Comment (benrudolph) [inline corehq/apps/mocha/README.md]: this is very similar to the instructions here: https://github.com/dimagi/commcare-hq#javascript-tests. what issues did you run into when following those? or do you just think they belong here?

File: corehq/apps/mocha/README.md [hunk 5]
```diff
@@ -1,6 +1,33 @@
...
+
+1. Install `grunt-mocha`
+
+```
+npm install grunt-mocha
```
Comment (benrudolph) [inline corehq/apps/mocha/README.md]: this should get installed when one runs `npm install` (along with all other deps, so i recommend using that)

**Comment by benrudolph** [comment]
this is awesome! thanks for following up on adding the tests. makes me feel much better about the maintainability of the code. i have one semi-blocking comment on the docs cause i think it's duplication, but other than that :+1:

File: corehq/apps/export/static/export/spec/list_exports.spec.mo.js (hunk 2 above)
Comment (biyeun) [inline corehq/apps/export/static/export/spec/list_exports.spec.mo.js]: yep that's the way to do it for angular js

**Comment by biyeun** [comment]
@benrudolph sorry, didn't know to look in both places. Perhaps a link in mocha's readme to the main docs for setup would make more sense

**Comment by czue** [comment]
this is awesome, thanks! i haven't had a chance to learn much about js testing, so will defer to @benrudolph on detailed review/merge

**Comment by benrudolph** [comment]
legit. i think a link could be useful. the reason it's in two places is that the README on HQ just tells you how to run the tests. where as the readme in the mocha directory actually tells you how to write a new mocha test suite. i thought the later was too detailed for our main README, and the former too necessary to be tucked away in the mocha app

**Comment by snopoke** [comment]
sweet tests

**Pull request merged**